### PR TITLE
Mlfow logging overhual PR (or mock PR)

### PR DESCRIPTION
**Pull request by wli51**
**Summary**

This is a **minimal demonstration PR** intended to outline the proposed design for a major overhaul of the `virtual_stain_flow` software. The goal is to enable **better integration of MLflow logging** into the virtual staining model training process. 

To avoid disrupting existing functionality, all new features are implemented in separate `*V2.py` files. These filenames are **tentative** and subject to change. If the proposed design is approved by other contributors, this PR may evolve into a finalized implementation (with further refinements), or be discarded in favor of an alternative design.

---

**Key Additions and Changes**

1. **Modularized MLflow Logger**  
   - The `MlflowLogger` class has been **extracted from the `callbacks` module** and promoted to a standalone module: `logging.MlflowLoggerV2`.

2. **New Trainer Class with Logging Integration**  
   - Introduces a new `FNetTrainerV2` in `trainers/TrainerV2.py`, which closely mirrors the original `FNetTrainer` in `trainers/Trainer.py`.
   - This new trainer accepts a `logging.MlflowLoggerV2` instance in the `.train()` call and invokes it throughout the training cycle to **automatically log all computed metrics to MLflow**.

3. **New Callback Submodule for MLflow Logging**  
   - Adds a `logging.callbacks` submodule, distinct from the existing `callbacks` module.
   - These callback classes are designed to be passed into `MlflowLoggerV2` instances and can:
     - Generate plot artifacts (`.png` files)
     - Return metrics as key-value pairs
     - Return parameters as dictionaries  
   - `MlflowLoggerV2` **automatically logs** these outputs as MLflow artifacts, metrics, and parameters.

---

**Planned but Not Yet Implemented (to keep PR minimal)**

1. Overhaul of the `datasets` module to support logging dataset parameters during training for **reproducibility** under the same MLflow run.
2. Addition of a **log exchange data class** to act as a structured interface between `MlflowLoggerV2` and `logging.callbacks`, supporting richer artifact and parameter logging.
3. Expansion of `MlflowLoggerV2` to **integrate with Optuna trials** for hyperparameter tuning support.
4. Update of the wGaN model and trainer to support the new logging system.

---

See `examples/prototype_logging.ipynb` and `logging/README.md` for details

### CODE REVIEW

**Review by MattsonCam — APPROVED** [review]
This LGTM @wli51, the explanations were helpful and I think this is a solid approach. Also, I think there are a few minor changes that can be made here (aside from the TODOs), but it looks like it coming together nicely, good job!

Tangentially, I noticed that you're name files with `*V2.py` to avoid disrupting existing functionality. One way to mitigate this problem in the future would be to create a development branch (dev), and then create other branches from dev, such as `dev-0.2-prototype-logger`. That way you are not disrupting existing functionality in main, and you can create smaller prs to be reviewed and then merged into dev. After you have finished development on dev, then somebody could perform one final review before the code is merged into main.

**Comment by MattsonCam** [comment]
Should also mention:
I think it would be useful to show the structure and interactions of classes in this project. Up to you, but something like a UML diagram could help users understand or remember how to use this tool.

File: logging/MlflowLoggerV2.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+                raise RuntimeError(f"Error setting MLflow tracking URI: {e}")
+
+        self.experiment_name = experiment_name
+        self.run_name = run_name
+        self.analysis_type = experiment_type
+        self.model_architecture = model_architecture
+        self.target_channel_name = target_channel_name
+
+        self.trainer: Optional[AbstractLoggingTrainer] = None
+        self.tags = tags or {}
```
[the remark this answers is not in the document]
Comment (wli51) [reply logging/MlflowLoggerV2.py]: I completely forgot about them. Nice catch! Fixed.

File: logging/MlflowLoggerV2.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        
+        if self.trainer is None:
+            # TODO consider if we want error out or return None
+            return None
+            # raise RuntimeError("No trainer bound to this logger. Cannot access model.")
+        
+        if best_model:
+            return self.trainer.best_model
+        else:
+            return self.trainer.model
```
[the remark this answers is not in the document]
Comment (wli51) [reply logging/MlflowLoggerV2.py:484]: i think this is the best way for now and will be the best way in the future in terms of support for expansion.  An additional source of complexity that could arise from trying to retrieve model from mlflow would be the handling of multiple networks, and having the model class trainers define model access would be the most convenient and intuitive. 

That said, this class and PR as it stands today does not handle any multi-network returns from the trainer. Multi-model handling by the logger and artifact logging will be added in the upcoming PR when we expand the new logging framework to wGaN and trainer where we will get 2 networks.

File: trainers/AbstractLoggingTrainer.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            patience=patience,
+            callbacks=callbacks,
+            metrics=metrics,
+            device=device,
+            early_termination_metric=early_termination_metric
+        )
+
+        if early_termination_mode not in ["min", "max"]:
+            raise ValueError("early_termination_mode must be either 'min' or 'max'")
+        self._early_termination_mode = early_termination_mode
```
[the remark this answers is not in the document]
Comment (wli51) [reply trainers/AbstractLoggingTrainer.py:57]: I think a max vs min optimization mode switch for the early termination metrics would be a good to have still. 

You are right in pointing out that we could consistently define metrics in such a way that minimal=good model performance. But image that someone wants to optimize their model against an image quality assessment metric like the PSNR (this is not the best example), and have to create a wrapper class `NegativePSNR`. This can be a somewhat unecessary burden had the max vs min switch was present. Moreover, the `NegativePSNR` gets logged as a metric in mlflow and we would be faced with the question of do we want our log to be populated by a transformed metric that is natural.

File: trainers/AbstractLoggingTrainer.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+- Binding of logger class moved to train method
+- Early termination mode is now a parameter of the class
+- The train method now also includes invocation of the logger.on_train_start(), ...,  logger.on_train_end() methods
+- Requires child classes to implement the save_model method
+"""
+class AbstractLoggingTrainer(AbstractTrainer):
+    """
+    Abstract class for trainers that support renovated mlflow logging functionality.
+    The new mlflow logger is no longer a callback but a separate class that can take
+    its own callbacks to faciliate centralized logging of artifacts.
```
[the remark this answers is not in the document]
Comment (wli51) [reply trainers/AbstractLoggingTrainer.py]: Moved!

File: logging/MlflowLoggerV2.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+        :param tag: The tag to associate with the logged output.
+        :param callback_return: The output from the callback, which can be:
+            - pathlib.Path: when a file is produced, log as an artifact
+            - Dict[str, Any]: when a dictionary is produced, log as a parameter
+            - Tuple[str, Any]: when a key value tuple is produced, log as a metric
+        :param stage: Optional stage to categorize the logged output, defaults to None.
+        :param step: Optional step or epoch at which the metric is logged, defaults to None.
+        """
+
+        if callback_return is None:
```
Comment (wli51) [inline logging/MlflowLoggerV2.py]: reduce complexity

File: logging/MlflowLoggerV2.py [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+    
+    """
+    Access point for callback to model
+    """
+    def get_epoch(
+            self
+        ) -> int:
+
+        if self.trainer is None:
+            # TODO consider if we want error out or return None
```
[the remark this answers is not in the document]
Comment (wli51) [reply logging/MlflowLoggerV2.py]: UML diagram to explain the topology and dependencies for users

File: logging/callbacks/PlotCallback.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        plot_file_path = f"{self._path}/epoch_{self.get_epoch()}.png"
+        plot_predictions_grid_from_model(
+            model=model,
+            dataset=self._dataset,
+            indices=self._dataset_subset_indices,
+            metrics=self.plot_metrics,
+            save_path=plot_file_path,
+            device=original_device,
+            **self.plot_kwargs
+        )
```
[the remark this answers is not in the document]
Comment (wli51) [reply logging/callbacks/PlotCallback.py:145]: plotly?
